common/lib/xmodule/xmodule/seq_module.py

```diff
@@ -312,55 +341,15 @@ def gate_sequence_if_it_is_a_timed_exam_and_contains_content_type_gated_problems
         We are displaying the gating fragment within the sequence, as is done for gating for prereqs,
         rather than content type gating the entire sequence because that would remove the next/previous navigation.
 
-        When gated_sequence_fragment is not set to None, the sequence will be gated.
+        When gated_sequence_paywall is not set to None, the sequence will be gated.
 
         This functionality still needs to be replicated in the frontend-app-learning courseware MFE
         The ticket to track this is https://openedx.atlassian.net/browse/REV-1220
         Note that this will break compatability with using sequences outside of edx-platform
         but we are ok with this for now
         """
-        if not self.is_time_limited:
-            self.gated_sequence_fragment = None
-            return
-
-        try:
-            user = self._get_user()
-            course_id = self.runtime.course_id
-            content_type_gating_service = self.runtime.service(self, 'content_type_gating')
-            if not (content_type_gating_service and
-                    content_type_gating_service.enabled_for_enrollment(user=user, course_key=course_id)):
-                self.gated_sequence_fragment = None
-                return
-
-            def leaf_filter(block):
-                # This function is used to check if this is a leaf block
-                # Blocks with children are not currently gated by content type gating
-                # Other than the outer function here
-                return (
-                    block.location.block_type not in ('chapter', 'sequential', 'vertical') and
-                    not block.has_children
-                )
-
-            def get_children(parent):
-                # This function is used to get the children of a block in the traversal below
-                if parent.has_children:
-                    return parent.get_children()
-                else:
-                    return []
-
-            # If any block inside a timed exam has been gated by content type gating
-            # then gate the entire sequence.
-            # In order to avoid scope creep, we are not handling other potential causes
-            # of access failures as part of this work.
-            for block in traverse_pre_order(self, get_children, leaf_filter):
-                gate_fragment = content_type_gating_service.content_type_gate_for_block(user, block, course_id)
-                if gate_fragment is not None:
-                    self.gated_sequence_fragment = gate_fragment
-                    return
-                else:
-                    self.gated_sequence_fragment = None
-        except User.DoesNotExist:
-            self.gated_sequence_fragment = None
+        if self.is_time_limited and self.gated_sequence_paywall:
+            return self.gated_sequence_paywall
```

Comment:
Does this not need to call `_check_children_for_content_type_gating_paywall`? I might just be getting turned around in the code, but it looks like `gate_entire_sequence_if_it_is_a_timed_exam_and_contains_content_type_gated_problems` is called before the check_children function which would mean this is (almost) always None, right?

Reply:
perhaps there is a clearer way to structure this, but self.gated_sequence_paywall is set to None in init
then _check_children_for_content_type_gating_paywall it is set in _render_student_view_for_items here https://github.com/edx/edx-platform/pull/26143/files#diff-20826fcfc64f6ff8931e4b2e686a9a5a588b59c2eb808bb0eee0ed3ae4efa575R453
and gate_entire_sequence_if_it_is_a_timed_exam_and_contains_content_type_gated_problems is called after that  

so i think it should be correctly called before that function call, but i'm open to refactoring suggestions to make it clearer

Reply:
Hmmm. So an easier solution could be to just add in the docstring that the other method is called first so that value should be set. A more difficult solution is to leave the init to None, but then change the return value in the children function to True or False (or Fragment or False) so that way this function can check if the value is still None (indicating the children function has not been called yet), it will call it and otherwise it will use the result it has. But that's obviously a bigger lift and not necessary, but would appreciate at least a warning in the docstring.

Reply:
i've added two comments for now.  
in theory you could have a self.function_has_been_called variable though that seems like it shouldn't be necessary

Reply:
Works for me!